File: tools/codex-relay-client/src/LanAi.RelayClient.WeChatReader.Mac/Sources/wechat-reader/TextRecognizer.swift
import CoreGraphics
import Foundation
import Vision

/// Vision's text recogniser, Simplified Chinese first. On the device; nothing is sent anywhere.
enum TextRecognizer {
    /// Whether this macOS offers Simplified Chinese recognition (macOS 13+ at the accurate level).
    static func supportsChinese() -> Bool {
        let request = VNRecognizeTextRequest()
        request.recognitionLevel = .accurate
        let languages = (try? request.supportedRecognitionLanguages()) ?? []
        return languages.contains("zh-Hans")
    }

    /// Recognises the region and returns its lines in the image's coordinates, top to bottom.
    static func read(_ p: Pixels, x: Int, y: Int, w: Int, h: Int, scale: Double) throws -> [ReaderLine] {
        let region = CGRect(x: max(0, x), y: max(0, y),
                            width: min(w, p.width - max(0, x)), height: min(h, p.height - max(0, y)))
        guard region.width > 0, region.height > 0, let crop = p.image.cropping(to: region) else { return [] }

        let request = VNRecognizeTextRequest()
        request.recognitionLevel = .accurate
        request.recognitionLanguages = ["zh-Hans", "en-US"]
        request.usesLanguageCorrection = true
        try VNImageRequestHandler(cgImage: crop, options: [:]).perform([request])

        let width = Double(region.width)
        let height = Double(region.height)
        var lines: [ReaderLine] = []
        for observation in request.results ?? [] {
            guard let candidate = observation.topCandidates(1).first else { continue }
            // Vision's boxes are normalised with the origin at the bottom left.
            let box = observation.boundingBox
            let lx = Int(region.minX) + Int(box.minX * width)
            let ly = Int(region.minY) + Int((1 - box.maxY) * height)
            let lw = Int(box.width * width)
            let lh = Int(box.height * height)
            lines.append(ReaderLine(text: candidate.string, x: lx, y: ly, w: lw, h: lh,
                                    bg: ChatLayout.sampleBehind(p, x: lx, y: ly, w: lw, h: lh, scale: scale)))
        }

        return lines.sorted { $0.y != $1.y ? $0.y < $1.y : $0.x < $1.x }
    }
}

File: tools/codex-relay-client/src/LanAi.RelayClient.WeChatReader/TextRecognizer.cs
using System.Runtime.InteropServices.WindowsRuntime;
using System.Text;
using LanAi.RelayClient.WeChatIntent;
using Windows.Globalization;
using Windows.Graphics.Imaging;
using Windows.Media.Ocr;

namespace LanAi.RelayClient.WeChatReader;

/// <summary>Windows' built-in OCR, Simplified Chinese.</summary>
/// <remarks>
/// Measured at 120–175 ms for a whole 827×961 window on the development machine; reading only
/// the message list is smaller still. Nothing is sent anywhere — the engine is part of Windows.
/// </remarks>
internal sealed class TextRecognizer
{
    private readonly OcrEngine _engine;

    private TextRecognizer(OcrEngine engine) => _engine = engine;

    /// <summary>Null when Windows has no Simplified Chinese recogniser installed.</summary>
    public static TextRecognizer? TryCreate()
    {
        OcrEngine? engine = OcrEngine.TryCreateFromLanguage(new Language("zh-Hans-CN"))
            ?? OcrEngine.TryCreateFromLanguage(new Language("zh-Hans"));
        return engine is null ? null : new TextRecognizer(engine);
    }

    /// <summary>
    /// Recognises the region <paramref name="x"/>,<paramref name="y"/>,<paramref name="w"/>,<paramref name="h"/>
    /// and returns its lines in the window's coordinates, top to bottom.
    /// </summary>
    public async Task<List<ReaderLine>> ReadAsync(Pixels p, int x, int y, int w, int h, double scale)
    {
        x = Math.Clamp(x, 0, p.Width - 1);
        y = Math.Clamp(y, 0, p.Height - 1);
        w = Math.Clamp(w, 1, p.Width - x);
        h = Math.Clamp(h, 1, p.Height - y);

        var crop = new byte[w * h * 4];
        for (int row = 0; row < h; row++)
        {
            Buffer.BlockCopy(p.Bgra, (((y + row) * p.Width) + x) * 4, crop, row * w * 4, w * 4);
        }

        using var bgra = new SoftwareBitmap(BitmapPixelFormat.Bgra8, w, h, BitmapAlphaMode.Premultiplied);
        bgra.CopyFromBuffer(crop.AsBuffer());
        using SoftwareBitmap gray = SoftwareBitmap.Convert(bgra, BitmapPixelFormat.Gray8);
        OcrResult result = await _engine.RecognizeAsync(gray);

        var lines = new List<ReaderLine>();
        foreach (OcrLine line in result.Lines)
        {
            if (line.Words.Count == 0)
            {
                continue;
            }

            int left = (int)line.Words.Min(word => word.BoundingRect.Left);
            int top = (int)line.Words.Min(word => word.BoundingRect.Top);
            int right = (int)Math.Ceiling(line.Words.Max(word => word.BoundingRect.Right));
            int bottom = (int)Math.Ceiling(line.Words.Max(word => word.BoundingRect.Bottom));
            int lx = x + left, ly = y + top, lw = right - left, lh = bottom - top;
            lines.Add(new ReaderLine
            {
                Text = Join(line.Words),
                X = lx,
                Y = ly,
                W = lw,
                H = lh,
                Bg = ChatLayout.SampleBehind(p, lx, ly, lw, lh, scale),
            });
        }

        lines.Sort((a, b) => a.Y != b.Y ? a.Y.CompareTo(b.Y) : a.X.CompareTo(b.X));
        return lines;
    }

    /// <summary>
    /// The engine splits Chinese into one "word" per character or run. Words are joined without
    /// spaces, except between two Latin letters or digits, where the space was real.
    /// </summary>
    private static string Join(IReadOnlyList<OcrWord> words)
    {
        var text = new StringBuilder();
        foreach (OcrWord word in words)
        {
            if (text.Length > 0 && IsLatin(text[^1]) && word.Text.Length > 0 && IsLatin(word.Text[0]))
            {
                text.Append(' ');
            }

            text.Append(word.Text);
        }

        return text.ToString();
    }

    private static bool IsLatin(char c) => c < 128 && char.IsLetterOrDigit(c);
}
